miner/worker.go

```diff
@@ -775,11 +785,234 @@ func (w *worker) updateSnapshot() {
 		w.current.state.RevertToSnapshot(snap)
 		return nil, err
 	}
+
+	proofAccounts := tracer.UpdatedAccounts()
+	if w.chainConfig.Scroll.FeeVaultEnabled() {
+		proofAccounts[*w.chainConfig.Scroll.FeeVaultAddress] = struct{}{}
+	} else {
+		proofAccounts[coinbase] = struct{}{}
+	}
+	proofAccounts[rcfg.L1GasPriceOracleAddress] = struct{}{}
+	for addr := range proofAccounts {
+		addrStr := addr.String()
+		if _, existed := txStorageTrace.Proofs[addrStr]; existed {
+			continue
+		}
+		proof, err := w.current.state.GetProof(addr)
+		if err != nil {
+			log.Error("Proof not available", "address", addrStr, "error", err)
+			// but we still mark the proofs map with nil array
+		}
+		wrappedProof := make([]hexutil.Bytes, len(proof))
+		for i, bt := range proof {
+			wrappedProof[i] = bt
+		}
+		txStorageTrace.Proofs[addrStr] = wrappedProof
+	}
+
+	proofStorages := tracer.UpdatedStorages()
+	proofStorages[rcfg.L1GasPriceOracleAddress] = vm.Storage(
+		map[common.Hash]common.Hash{
+			rcfg.L1BaseFeeSlot: {},
+			rcfg.OverheadSlot:  {},
+			rcfg.ScalarSlot:    {},
+		})
+	for addr, keys := range proofStorages {
+		if _, existed := txStorageTrace.StorageProofs[addr.String()]; !existed {
+			txStorageTrace.StorageProofs[addr.String()] = make(map[string][]hexutil.Bytes)
+		}
+
+		trie, err := w.current.state.GetStorageTrieForProof(addr)
+		if err != nil {
+			// but we still continue to next address
+			log.Error("Storage trie not available", "error", err, "address", addr)
+			continue
+		}
+		zktrieTracer := w.current.state.NewProofTracer(trie)
+
+		for key, values := range keys {
+			addrStr := addr.String()
+			keyStr := key.String()
+			isDelete := bytes.Equal(values.Bytes(), common.Hash{}.Bytes())
+
+			m, existed := txStorageTrace.StorageProofs[addrStr]
+			if !existed {
+				m = make(map[string][]hexutil.Bytes)
+				txStorageTrace.StorageProofs[addrStr] = m
+				if zktrieTracer.Available() {
+					zkTrieTracers[addrStr] = zktrieTracer
+				}
+			} else if _, existed := m[keyStr]; existed {
+				// still need to touch tracer for deletion
+				if isDelete && zktrieTracer.Available() {
+					zkTrieTracers[addrStr].MarkDeletion(key)
+				}
+				continue
+			}
+
+			var proof [][]byte
+			var err error
+			if zktrieTracer.Available() {
+				proof, err = w.current.state.GetSecureTrieProof(zktrieTracer, key)
+			} else {
+				proof, err = w.current.state.GetSecureTrieProof(trie, key)
+			}
+			if err != nil {
+				log.Error("Storage proof not available", "error", err, "address", addrStr, "key", keyStr)
+				// but we still mark the proofs map with nil array
+			}
+			wrappedProof := make([]hexutil.Bytes, len(proof))
+			for i, bt := range proof {
+				wrappedProof[i] = bt
+			}
+			m[keyStr] = wrappedProof
+			if zktrieTracer.Available() {
+				if isDelete {
+					zktrieTracer.MarkDeletion(key)
+				}
+				zkTrieTracers[addrStr].Merge(zktrieTracer)
+			}
+		}
+	}
+
+	// build dummy per-tx deletion proof
+	for _, zkTrieTracer := range zkTrieTracers {
```

Comment:
The "per-tx" deletion proof is not enough since we in fact need "deletion proofs from the first tx till current one".

We should keep the per-account `zktrieTracer` map in the context of miner object and only reset it in `makeCurrent`. (If we care about effecient, we can also kept the per-account storage trie obtained from `GetStorageTrieForProof`)